File: python/src/dashgps/formats/nmea.py
"""NMEA 0183 sentences embedded in TS private streams or MP4 atoms.

Implements spec/04-nmea.md. NMEA is a published standard, so this parser is the catch-all that
gives dashgps a chance on cameras nobody has documented.

Status: UNTESTED against a real dashcam; the wire format itself is standard.
"""

from ..containers.mp4 import iter_atoms, looks_like_mp4
from ..containers.ts import WELL_KNOWN_PIDS, detect_alignment, iter_pes
from ..fmt import NAN, epoch_from_civil, parse_num
from ..io import capped_windows, scan_chunks
from ..model import ParseError, ParseResult, Point

FORMAT_ID = "nmea"
STATUS = "untested"

ATOMS = (b"free", b"skip", b"udta", b"gps ")
KNOTS_TO_KMH = 1.852


def _checksum_ok(body, given):
    x = 0
    for b in body:
        x ^= b
    hi = "0123456789ABCDEF"[(x >> 4) & 15]
    lo = "0123456789ABCDEF"[x & 15]
    return given.upper() == hi + lo


def iter_sentences(buf):
    """Yield (talker_and_type, fields, ok) for every $-delimited candidate."""
    i = 0
    n = len(buf)
    while True:
        s = buf.find(b"$", i)
        if s < 0:
            return
        e = s + 1
        while e < n and buf[e] not in (0x24, 0x00, 0x0D, 0x0A):
            e += 1
        raw = buf[s + 1 : e]
        i = e if e > s else s + 1
        if len(raw) < 7:
            continue
        star = raw.rfind(b"*")
        ok = True
        if star >= 0 and len(raw) - star >= 3:
            ok = _checksum_ok(raw[:star], raw[star + 1 : star + 3].decode("ascii", "replace"))
            raw = raw[:star]
        try:
            text = raw.decode("ascii")
        except UnicodeDecodeError:
            continue
        parts = text.split(",")
        if not parts or len(parts[0]) < 5:
            continue
        yield parts[0], parts, ok


def _dm_to_deg(v, hemi, pos):
    if v != v:
        return NAN
    deg = int(v / 100.0)
    out = deg + (v - deg * 100.0) / 60.0
    return out if hemi == pos else -out


def _rmc(parts):
    # time, status, lat, NS, lon, EW, knots, track, date, magvar, magvarEW
    if len(parts) < 10 or parts[2] != "A":
        return None
    tm = parts[1]
    dt = parts[9]
    if len(tm) < 6 or len(dt) != 6:
        return None
    try:
        hh = int(tm[0:2])
        mi = int(tm[2:4])
        ss = int(tm[4:6])
        dd = int(dt[0:2])
        mo = int(dt[2:4])
        yy = int(dt[4:6])
    except ValueError:
        return None
    year = 1900 + yy if yy >= 70 else 2000 + yy
    if mo < 1 or mo > 12 or dd < 1 or dd > 31 or hh > 23 or mi > 59 or ss > 60:
        return None
    lat = _dm_to_deg(parse_num(parts[3]), parts[4], "N")
    lon = _dm_to_deg(parse_num(parts[5]), parts[6], "E")
    if lat != lat or lon != lon:
        return None
    kn = parse_num(parts[7])
    magvar = NAN
    if len(parts) >= 12:
        mv = parse_num(parts[10])
        if mv == mv:
            magvar = mv if parts[11] == "E" else -mv
    return {
        "key": tm,
        "t": epoch_from_civil(year, mo, dd, hh, mi, 59 if ss == 60 else ss),
        "lat": lat, "lon": lon,
        "speed": kn * KNOTS_TO_KMH if kn == kn else NAN,
        "track": parse_num(parts[8]),
        "magvar": magvar,
    }


def _gga(parts):
    if len(parts) < 10:
        return None
    return {"key": parts[1], "alt": parse_num(parts[9])}


def _buffers(reader, opts):
    """Yield byte buffers that might contain NMEA, from whichever container this is."""
    if looks_like_mp4(reader):
        cap = opts.tail_cap if opts.tail_cap > 0 else 1024 * 1024
        for a in iter_atoms(reader):
            if a.type in ATOMS and 0 < a.body_size <= cap:
                yield reader.read_range(a.body, a.end)
        return
    stride, off = detect_alignment(reader)
    if stride:
        end = reader.size()
        if 0 < opts.scan_cap < end:
            end = opts.scan_cap
        for _pid, payload in iter_pes(reader, stride, off, WELL_KNOWN_PIDS, opts.chunk, end=end):
            yield payload
        return
    if opts.raw_nmea:
        for start, stop in capped_windows(reader.size(), opts.scan_cap):
            for slab, _first in scan_chunks(reader, opts.chunk, opts.overlap, start, stop):
                yield slab.data


def _collect(reader, opts, res, limit=0):
    pend_rmc = None
    pend_gga = None
    pend_key = None
    rejects = 0
    seen = 0
    out = []

    def flush():
        if pend_rmc is None:
            return
        alt = pend_gga["alt"] if pend_gga is not None else NAN
        out.append(Point(pend_rmc["t"] - 0, pend_rmc["lat"], pend_rmc["lon"],
                         pend_rmc["speed"], pend_rmc["track"], alt, pend_rmc["magvar"],
                         NAN, NAN, NAN, -1, 0))

    for buf in _buffers(reader, opts):
        for kind, parts, ok in iter_sentences(buf):
            t = kind[2:5]
            if t not in ("RMC", "GGA"):
                continue
            if not ok:
                rejects += 1
                continue
            seen += 1
            key = parts[1] if len(parts) > 1 else ""
            # Pairing is on the NMEA time field, not on line order. spec 04.
            if pend_key is not None and key != pend_key:
                flush()
                pend_rmc = None
                pend_gga = None
            pend_key = key
            if t == "RMC":
                r = _rmc(parts)
                if r is not None:
                    pend_rmc = r
            else:
                g = _gga(parts)
                if g is not None:
                    pend_gga = g
            if limit and len(out) >= limit:
                return out, seen, rejects
    flush()
    return out, seen, rejects


def sniff(reader, opts):
    if not opts.deep:
        return 0.0
    pts, seen, _ = _collect(reader, opts, ParseResult(FORMAT_ID, STATUS), limit=1)
    if pts:
        return 0.7
    return 0.0


def parse(reader, opts):
    res = ParseResult(FORMAT_ID, STATUS, sources=[reader.name], time_is_naive=False)
    pts, seen, rejects = _collect(reader, opts, res)
    res.meta["sentences"] = seen
    res.meta["checksum_rejects"] = rejects
    if rejects:
        res.warn("%d NMEA sentences failed their checksum and were ignored" % rejects)
    if not pts:
        if seen:
            # GGA carries no date, so a GGA-only stream cannot produce absolute timestamps.
            raise ParseError("NMEA sentences found but none carried a valid RMC fix")
        raise ParseError("no NMEA sentences found")
    res.points = pts
    return res

File: js/src/formats/nmea.js
// NMEA 0183 sentences in TS private streams or MP4 atoms. spec/04-nmea.md
// Mirrors python/src/dashgps/formats/nmea.py

import { iterAtoms, looksLikeMp4 } from "../containers/mp4.js";
import { WELL_KNOWN_PIDS, detectAlignment, iterPes } from "../containers/ts.js";
import { cappedWindows, scanChunks } from "../io.js";
import { epochFromCivil, parseNum } from "../fmt.js";
import { ParseError, ParseResult, Point } from "../model.js";

export const FORMAT_ID = "nmea";
export const STATUS = "untested";

const ATOMS = new Set(["free", "skip", "udta", "gps "]);
const KNOTS_TO_KMH = 1.852;
const HEX = "0123456789ABCDEF";

function checksumOk(buf, from, to, given) {
  let x = 0;
  for (let i = from; i < to; i++) x ^= buf[i];
  return given === HEX[(x >> 4) & 15] + HEX[x & 15];
}

export function* iterSentences(buf) {
  let i = 0;
  const n = buf.length;
  for (;;) {
    let s = -1;
    for (let k = i; k < n; k++) { if (buf[k] === 0x24) { s = k; break; } }
    if (s < 0) return;
    let e = s + 1;
    while (e < n && buf[e] !== 0x24 && buf[e] !== 0x00 && buf[e] !== 0x0d && buf[e] !== 0x0a) e++;
    i = e > s ? e : s + 1;
    let end = e;
    let ok = true;
    let star = -1;
    for (let k = e - 1; k > s; k--) { if (buf[k] === 0x2a) { star = k; break; } }
    if (star >= 0 && e - star >= 3) {
      ok = checksumOk(buf, s + 1, star,
        String.fromCharCode(buf[star + 1], buf[star + 2]).toUpperCase());
      end = star;
    }
    if (end - s - 1 < 6) continue;
    let text = "";
    for (let k = s + 1; k < end; k++) text += String.fromCharCode(buf[k]);
    const parts = text.split(",");
    if (!parts.length || parts[0].length < 5) continue;
    yield [parts[0], parts, ok];
  }
}

function dmToDeg(v, hemi, pos) {
  if (v !== v) return NaN;
  const deg = Math.trunc(v / 100);
  const out = deg + (v - deg * 100) / 60;
  return hemi === pos ? out : -out;
}

function rmc(parts) {
  if (parts.length < 10 || parts[2] !== "A") return null;
  const tm = parts[1];
  const dt = parts[9];
  if (tm.length < 6 || dt.length !== 6) return null;
  const hh = Number(tm.slice(0, 2)), mi = Number(tm.slice(2, 4)), ss = Number(tm.slice(4, 6));
  const dd = Number(dt.slice(0, 2)), mo = Number(dt.slice(2, 4)), yy = Number(dt.slice(4, 6));
  if (![hh, mi, ss, dd, mo, yy].every(Number.isInteger)) return null;
  const year = yy >= 70 ? 1900 + yy : 2000 + yy;
  if (mo < 1 || mo > 12 || dd < 1 || dd > 31 || hh > 23 || mi > 59 || ss > 60) return null;
  const lat = dmToDeg(parseNum(parts[3]), parts[4], "N");
  const lon = dmToDeg(parseNum(parts[5]), parts[6], "E");
  if (lat !== lat || lon !== lon) return null;
  const kn = parseNum(parts[7]);
  let magvar = NaN;
  if (parts.length >= 12) {
    const mv = parseNum(parts[10]);
    if (mv === mv) magvar = parts[11] === "E" ? mv : -mv;
  }
  return {
    key: tm,
    t: epochFromCivil(year, mo, dd, hh, mi, ss === 60 ? 59 : ss),
    lat, lon,
    speed: kn === kn ? kn * KNOTS_TO_KMH : NaN,
    track: parseNum(parts[8]),
    magvar,
  };
}

function gga(parts) {
  if (parts.length < 10) return null;
  return { key: parts[1], alt: parseNum(parts[9]) };
}

async function* buffers(reader, opts) {
  if (await looksLikeMp4(reader)) {
    const cap = opts.tailCap > 0 ? opts.tailCap : 1024 * 1024;
    for await (const a of iterAtoms(reader)) {
      if (ATOMS.has(a.type) && a.bodySize > 0 && a.bodySize <= cap) {
        yield await reader.readRange(a.body, a.end);
      }
    }
    return;
  }
  const [stride, off] = await detectAlignment(reader);
  if (stride) {
    let end = reader.size();
    if (opts.scanCap > 0 && opts.scanCap < end) end = opts.scanCap;
    for await (const [, payload] of
        iterPes(reader, stride, off, WELL_KNOWN_PIDS, opts.chunk, true, 65536, end)) {
      yield payload;
    }
    return;
  }
  if (opts.rawNmea) {
    for (const [start, stop] of cappedWindows(reader.size(), opts.scanCap)) {
      for await (const [slab] of scanChunks(reader, opts.chunk, opts.overlap, start, stop)) {
        yield slab.data;
      }
    }
  }
}

async function collect(reader, opts, limit = 0) {
  let pendRmc = null;
  let pendGga = null;
  let pendKey = null;
  let rejects = 0;
  let seen = 0;
  const out = [];
  const flush = () => {
    if (pendRmc === null) return;
    const alt = pendGga !== null ? pendGga.alt : NaN;
    out.push(new Point(pendRmc.t, pendRmc.lat, pendRmc.lon, pendRmc.speed, pendRmc.track,
      alt, pendRmc.magvar, NaN, NaN, NaN, -1, 0));
  };
  for await (const buf of buffers(reader, opts)) {
    for (const [kind, parts, ok] of iterSentences(buf)) {
      const t = kind.slice(2, 5);
      if (t !== "RMC" && t !== "GGA") continue;
      if (!ok) { rejects += 1; continue; }
      seen += 1;
      const key = parts.length > 1 ? parts[1] : "";
      // Pairing is on the NMEA time field, not on line order. spec 04
      if (pendKey !== null && key !== pendKey) {
        flush();
        pendRmc = null;
        pendGga = null;
      }
      pendKey = key;
      if (t === "RMC") {
        const r = rmc(parts);
        if (r !== null) pendRmc = r;
      } else {
        const g = gga(parts);
        if (g !== null) pendGga = g;
      }
      if (limit && out.length >= limit) return [out, seen, rejects];
    }
  }
  flush();
  return [out, seen, rejects];
}

export async function sniff(reader, opts) {
  if (!opts.deep) return 0.0;
  const [pts] = await collect(reader, opts, 1);
  return pts.length ? 0.7 : 0.0;
}

export async function parse(reader, opts) {
  const res = new ParseResult(FORMAT_ID, STATUS, [reader.name], [], {}, [], 0, false);
  const [pts, seen, rejects] = await collect(reader, opts);
  res.meta.sentences = seen;
  res.meta.checksum_rejects = rejects;
  if (rejects) res.warn(rejects + " NMEA sentences failed their checksum and were ignored");
  if (!pts.length) {
    if (seen) throw new ParseError("NMEA sentences found but none carried a valid RMC fix");
    throw new ParseError("no NMEA sentences found");
  }
  res.points = pts;
  return res;
}
